# 에러
가끔 우리가 파이썬 코드를 만들며 에러를 범할 수 있습니다. 만일 버그(문제가 되는 코드)가 우리 프로그램에 존재한다면 우리는 그것을 제거하거나 무시할 수 있습니다.

# 에러 제거
에러를 제거하려면 문제가 되는 코드를 문제가 되지 않게 바꾸어야 합니다. 파이썬에서는 우리가 코드를 실행하면 어디에서 문제가 있는지 알려 줍니다.

# 에러 무시와 예외 처리
예를 들어서 만일 우리가 파이썬을 이용하여 게슨기를 만들고 싶다고 가정을 해봅시다. 그렇다면 만일 사용자가 숫자가 아닌 문자를 입력하면 어떻게 해야할까요? 우리는 여기에서 예외 처리를 할 수 있습니다.

예외 처리는 try를 사용하여 그 키워드 안에 있는 함수들에서 에러가 발생하면 무시 할 수 있습니다.
```
try:
  ur = int(input("2로 나눌 수를 입력하세요: "))
  print(ur)
except:
  print("버그! 버그! 버그! 🐛🐛🐛x999")
```

어떤 지정된 에러가 나면 실행하고 싶은 코드가 있다면 'except' 다음에 에러의 이름을 넣으면 됩니다.

```
try:
  ur = int(input("2로 나눌 수를 입력하세요: "))
  print(ur)
except ZeroDivisionError:
  print("버그! 버그! 버그! 🐛🐛🐛x999. 수를 0으로 나눌 수 없습니다. 🐛🐛🐛")
```

만일 try 안에 있는 코드에 아무런 문제가 없다면 'else'를 사용하여 어떤 코드를 실행할 수 있습니다.

```
try:
  ur = int(input("2로 나눌 수를 입력하세요: "))
  print(ur)
except ZeroDivisionError:
  print("버그! 버그! 버그! 🐛🐛🐛x999. 수를 0으로 나눌 수 없습니다. 🐛🐛🐛")
else:
  print('0이 아닙니다')
```

else와 비슷하지만 에러가 나도 코드를 실행하는 'finally'라는 키워드를 사용할 수 있습니다.

```
try:
  ur = int(input("2로 나눌 수를 입력하세요: "))
  print(ur)
except ZeroDivisionError:
  print("버그! 버그! 버그! 🐛🐛🐛x999. 수를 0으로 나눌 수 없습니다. 🐛🐛🐛")
finally:
  print('코드 끝!')
```

# 나만의 에러
파이썬에서 우리만의 에러를 만들 수 있습니다. 바로 'rais Exception()'이라는 메서드를 이용해서요. 괄호 안에 에러 설명을 넣을 수 있습니다.

```
x = -1

if x < 0:
  raise Exception("흠... 🐛가 존재합니다. X 변수가 0보다 작아요.")
```

지정된 에러 종류를 만들고 싶다면 'Exception' 대신 에러의 이름을 넣으면 됩니다.


